File: sql-chatbot-rails/lib/sql_chatbot/grammar/sanity_check.rb
# frozen_string_literal: true

module SqlChatbot
  module Grammar
    # Post-execution sanity check for grammar-generated SQL.
    #
    # Catches "plausible but wrong" results where the SQL ran without error
    # but the value disagrees with the registry's known row count. Concrete
    # case: Gitea's `user` reserved-word table returned 1 instead of 9.
    module SanityCheck
      # Returns { ok: true } when result matches expectations,
      # or { ok: false, reason: "..." } when there's a mismatch.
      def self.check_count(primitive, entity, result_rows)
        return { ok: true } unless primitive.to_s == "COUNT"
        return { ok: true } unless result_rows.is_a?(Array) && result_rows.length == 1

        row = result_rows.first
        v = row.is_a?(Hash) ? row.values.first : nil
        got = Integer(v.to_s) rescue nil
        return { ok: true } unless got

        expected = entity.row_count.to_i
        # Trust the registry only when it has a non-trivial value.
        # Tables with reltuples == 0 might just be stats-stale.
        return { ok: true } if expected <= 5

        if got < expected / 3 || got > expected * 3
          return {
            ok: false,
            reason: "count_mismatch: SQL returned #{got}, registry has ~#{expected} rows in #{entity.table}",
          }
        end
        { ok: true }
      end
    end
  end
end

File: packages/agent/src/grammar/sanity-check.ts
// Post-execution sanity check for grammar-generated SQL.
//
// Catches "plausible but wrong" results where the SQL ran without error but
// the value is way off. Concrete case: Gitea's reserved-word `user` table
// returned 1 instead of 9 because PG silently resolved `FROM user` to the
// CURRENT_USER function. Even with quoted identifiers (V1.2 #1), other
// classes of bug can produce numeric results that disagree with the
// registry's known row count — this catches them universally.

import { Entity } from './registry.js';

export interface SanityCheckResult {
  ok: boolean;
  reason?: string;
}

// Mismatch tolerance: known > 5 rows AND result is <1/3 or >3x the known count.
// The thresholds are tuned to avoid false positives on small tables (where
// the registry's pg_stat snapshot can be 0 due to no autovacuum yet).
export function checkCountSanity(
  primitive: string,
  entity: Entity,
  resultRows: unknown[],
): SanityCheckResult {
  if (primitive !== 'COUNT') return { ok: true };
  if (!Array.isArray(resultRows) || resultRows.length !== 1) return { ok: true };

  const row = resultRows[0] as Record<string, unknown>;
  const v = Object.values(row)[0];
  const got = Number(v);
  if (!Number.isFinite(got)) return { ok: true };

  const expected = entity.rowCount;
  // Trust the registry only when it has a non-trivial value.
  // Tables with reltuples == 0 might just be stats-stale, not empty.
  if (expected <= 5) return { ok: true };

  if (got < expected / 3 || got > expected * 3) {
    return {
      ok: false,
      reason: `count_mismatch: SQL returned ${got}, registry has ~${expected} rows in ${entity.table}`,
    };
  }
  return { ok: true };
}
